source/connection_manager.c

```diff
@@ -460,16 +487,127 @@ static MQTTStatus_t prvMQTTConnect( bool xCleanSession )
+     * command to the command queue. */
+    for( lIndex = 0; lIndex < SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS; lIndex++ )
+    {
+        /* Check if there is a subscription in the subscription list. This demo
+         * doesn't subscribe to the same topics from different tasks. If there
+         * are duplicates, duplicate topics should be avoided while sending
+         * resubscribes. */
+        if( xGlobalSubscriptionList[ lIndex ].usFilterStringLength != 0 )
+        {
+            xSubInfo[ usNumSubscriptions ].pTopicFilter = xGlobalSubscriptionList[ lIndex ].pcSubscriptionFilterString;
+            xSubInfo[ usNumSubscriptions ].topicFilterLength = xGlobalSubscriptionList[ lIndex ].usFilterStringLength;
+
+            /* QoS1 is used for all the subscriptions in this demo. */
+            xSubInfo[ usNumSubscriptions ].qos = MQTTQoS1;
+
+            usNumSubscriptions++;
+
+            LogInfo( ( "Resubscribe to the topic %.*s will be attempted.",
+                       xSubInfo[ usNumSubscriptions ].topicFilterLength,
+                       xSubInfo[ usNumSubscriptions ].pTopicFilter ) );
+        }
+    }
+
+    if( usNumSubscriptions > 0 )
+    {
+        xSubArgs.pSubscribeInfo = xSubInfo;
+        xSubArgs.numSubscriptions = usNumSubscriptions;
+
+        /* The block time can be 0 as the command loop is not running at this point. */
+        xCommandParams.blockTimeMs = 0U;
+        xCommandParams.cmdCompleteCallback = prvSubscriptionCommandCallback;
+        xCommandParams.pCmdCompleteCallbackContext = ( void * ) &xSubArgs;
+
+        /* Enqueue subscribe to the command queue. These commands will be processed only
+         * when command loop starts. */
+        xResult = MQTTAgent_Subscribe( &xGlobalMqttAgentContext, &xSubArgs, &xCommandParams );
+    }
+    else
+    {
+        /* Mark the resubscribe as success if there is nothing to be subscribed. */
+        xResult = MQTTSuccess;
+    }
+
+    if( xResult != MQTTSuccess )
+    {
+        LogError( ( "Failed to enqueue the MQTT subscribe command. xResult=%s.",
+                    MQTT_Status_strerror( xResult ) ) );
+    }
+
+    return xResult;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvSubscriptionCommandCallback( void * pxCommandContext,
+                                            MQTTAgentReturnInfo_t * pxReturnInfo )
+{
+    size_t lIndex = 0;
+    MQTTAgentSubscribeArgs_t * pxSubscribeArgs = ( MQTTAgentSubscribeArgs_t * ) pxCommandContext;
+
+    /* If the return code is success, no further action is required as all the topic filters
+     * are already part of the subscription list. */
+    if( pxReturnInfo->returnCode != MQTTSuccess )
+    {
+        /* Check through each of the suback codes and determine if there are any failures. */
+        for( lIndex = 0; lIndex < pxSubscribeArgs->numSubscriptions; lIndex++ )
+        {
+            /* If this is a suback failure. An attempt to resubscribe can be done.
```

Comment:
Grammar/confusing language: I'd simply say that this demo doesn't attempt to resubscribe in the event that a SUBACK failed, and delete the _"An attempt to resubscribe can be done."_